interfaces/builtin/fwupd.go

```diff
@@ -0,0 +1,229 @@
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package builtin
+
+import (
+	"bytes"
+	"github.com/snapcore/snapd/interfaces"
+)
+
+var fwupdPermanentSlotAppArmor = []byte(`
+# Description: Allow operating as the fwupd service. Reserved because this
+#  gives privileged access to the system.
+# Usage: reserved
+
+  # Allow read/write access for old efivars sysfs interface
+  capability sys_admin,
+
+  # File accesses
+  # Allow access for EFI System Resource Table in the UEFI 2.5+ specification
+  /sys/firmware/efi/esrt/entries/ r,
+  /sys/firmware/efi/esrt/entries/** r,
+
+  # Allow fwupd to access system information
+  /sys/devices/virtual/dmi/id/product_name r,
+  /sys/devices/virtual/dmi/id/sys_vendor r,
+
+  # File accesses
+  # Allow read access for efivarfs filesystem
+  /sys/firmware/efi/efivars/ r,
+  /sys/firmware/efi/efivars/** r,
+
+  # DBus accesses
+  #include <abstractions/dbus-strict>
+  dbus (send)
+      bus=system
+      path=/org/freedesktop/DBus
+      interface=org.freedesktop.DBus
+      member={Request,Release}Name
+      peer=(name=org.freedesktop.DBus),
+
+  dbus (receive, send)
+      bus=system
+      path=/org/freedesktop/DBus
+      interface=org.freedesktop.DBus
+      member=GetConnectionUnixUser
+      peer=(label=unconfined),
+
+  # Allow binding the service to the requested connection name
+  dbus (bind)
+      bus=system
+      name="org.freedesktop.fwupd",
+`)
+
+var fwupdConnectedPlugAppArmor = []byte(`
+# Description: Allow using fwupd service. Reserved because this gives
+#  privileged access to the fwupd service.
+# Usage: reserved
+
+  # DBus accesses
+  #include <abstractions/dbus-strict>
+
+  # Allow access to fwupd service
+  dbus (receive, send)
+      bus=system
+      path=/
+      interface=org.freedesktop.fwupd
+      peer=(label=###SLOT_SECURITY_TAGS###),
+
+  dbus (receive, send)
+      bus=system
+      path=/
+      interface=org.freedesktop.DBus.Properties
+      peer=(label=###SLOT_SECURITY_TAGS###),
+`)
+
+var fwupdConnectedSlotAppArmor = []byte(`
+# Description: Allow firmware update from fwupd service. Reserved because this gives
+#  privileged access to the system.
+# Usage: reserved
+
+  # Allow libfwup to access efivarfs with immutable flag
+  capability linux_immutable,
+
+  # File accesses
+  # Allow write access for efivarfs filesystem
+  /sys/firmware/efi/efivars/** rw,
+
+  # Allow write access for efi firmware updater
+  /boot/efi/EFI/ubuntu/fw/** rw,
+  # Allow access from efivar library
+  @{PROC}/@{pid}/mounts r,
+  /sys/devices/**/partition r,
+  /dev/sd[a-z]* r,
```

Comment:
efivar lib need full disk access to get the partition guid and size info, I made a patch at [https://github.com/timchen119/efivar/commit/331e4102249987e93e88185ae9e3103b5a2b9404]() as a proof of concept to use udev runtime data to get these info, so we may change this to use this following rules without open the full disk access:
```
/run/udev/data/b* r,
```
However honestly I'm looking for other suggestions as this is really a quick patch which need much more time to polish.